[Coverage][Single] Enable Branch coverage for loop statements

chapuni opened the pull request:
Depends on: #112730 #113114

https://discourse.llvm.org/t/rfc-integrating-singlebytecoverage-with-branch-coverage/82492

chapuni committed:
[Coverage] Move SingleByteCoverage out of CountedRegion

`SingleByteCoverage` is not per-region attribute at least.
At the moment, this change moves it into `FunctionRecord`.

chapuni committed:
[Coverage] Make SingleByteCoverage work consistent to merging

- Round `Counts` as 1/0
- Confirm both `ExecutionCount` and `AltExecutionCount` are in range.

chapuni committed:
Rework. (Also reverts  "[Coverage] Move SingleByteCoverage out of CountedRegion")

chapuni committed:
[Coverage] Introduce `getBranchCounterPair()`. NFC.

This aggregates the generation of branch counter pair as `ExecCnt` and
`SkipCnt`, to aggregate `CounterExpr::subtract`. At the moment:

- This change preserves the behavior of
  `llvm::EnableSingleByteCoverage`. Almost of SingleByteCoverage will
  be cleaned up by coming commits.

- `getBranchCounterPair()` is not called in
  `llvm::EnableSingleByteCoverage`. I will implement the new behavior
  of SingleByteCoverage in it.

- `IsCounterEqual(Out, Par)` is introduced instead of
  `Counter::operator==`. Tweaks would be required for the comparison
  for additional counters.

- Braces around `assert()` is intentional. I will add a statement there.

https://discourse.llvm.org/t/rfc-integrating-singlebytecoverage-with-branch-coverage/82492

chapuni committed:
Introduce the type `CounterPair` for RegionCounterMap

`CounterPair` can hold `<uint32_t, uint32_t>` instead of current
`unsigned`, to hold also the counter number of SkipPath. For now, this
change provides the skeleton and only `CounterPair::first` is used.

Each counter number can have `None` to suppress emitting counter
increment. `second` is initialized as `None` by default, since most
`Stmt*` don't have a pair of counters.

This change also provides stubs for the verifyer. I'll provide the
impl of verifier for `+Asserts` later.

`markStmtAsUsed(bool, Stmt*)` may be used to inform that other side
counter may not emitted.

`markStmtMaybeUsed(S)` may be used for the `Stmt` and its inner will
be excluded for emission in the case of skipping by constant
folding. I put it into places where I found.

`verifyCounterMap()` will check the coverage map the counter map and
can be used to report inconsistency.

These verifier methods shall be eliminated in `-Asserts`.

https://discourse.llvm.org/t/rfc-integrating-singlebytecoverage-with-branch-coverage/82492

chapuni committed:
Introduce CounterExpressionBuilder::replace(C, Map)

This return a counter for each term in the expression replaced by
ReplaceMap.

At the moment, this doesn't update the Map, so Map is marked as `const`.

chapuni committed:
[Coverage] Make additional counters available for BranchRegion. NFC.

`getBranchCounterPair()` allocates an additional Counter to SkipPath
in `SingleByteCoverage`.

`IsCounterEqual()` calculates the comparison with rewinding counter
replacements.

`NumRegionCounters` is updated to take additional counters in account.

`incrementProfileCounter()` has a few additiona arguments.

- `UseSkipPath=true`, to specify setting counters for SkipPath. It
  assumes `UseSkipPath=false` is used together.

- `UseBoth` may be specified for marking another path. It introduces
  the same effect as issueing `markStmtAsUsed(!SkipPath, S)`.

`llvm-cov` discovers counters in `FalseCount` to allocate
`MaxCounterID` for empty profile data.

chapuni committed:
Merge branches 'users/chapuni/cov/single/getpair', 'users/chapuni/cov/single/replace' and 'users/chapuni/cov/single/pair' into HEAD

chapuni committed:
clang/test/CoverageMapping/single-byte-counters.cpp: Rewrite counter matches

chapuni committed:
[Coverage][Single] Enable Branch coverage for loop statements

chapuni committed:
Rewind changes for folding

chapuni committed:
`first` may be cancelled.

Currently `first` is not None by default.

chapuni committed:
Update comments

chapuni committed:
Use initializer statements

chapuni committed:
Merge branches 'users/chapuni/cov/single/getpair' and 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
[Coverage][Single] Enable Branch coverage for CondOp

chapuni committed:
Merge branches 'users/chapuni/cov/single/test' and 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
CGF::markStmtAsUsed

chapuni committed:
CGF.markStmtMaybeUsed for binop

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

chapuni committed:
Introduce skeleton getSwitchImplicitDefaultCounter()

chapuni committed:
Update getSwitchImplicitDefaultCounter

chapuni committed:
Merge branches 'users/chapuni/cov/single/pair' and 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into HEAD

chapuni committed:
Don't allocate second if SkipExpr isn't Expr.

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
update

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

chapuni committed:
Suppress StepV in EnableSingleByteCoverage

chapuni committed:
assert(!StepV)

chapuni committed:
test/llvm-cov: Transform %.c* tests to {%.test, Inputs/%.c*}

And reformat. NFC.

chapuni committed:
Introduce test/llvm-cov/Inputs/yaml.makefile for convenience.

chapuni committed:
Add tests for SingleByteCoverage

chapuni committed:
Merge branch 'users/chapuni/cov/single/test' into users/chapuni/cov/single/merge

chapuni committed:
Fix a test to fix linecount=1

chapuni committed:
Merge branch 'users/chapuni/cov/single/merge' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/merge' into users/chapuni/cov/single/loop

chapuni committed:
Update single tests

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

chapuni committed:
Update single tests

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

chapuni committed:
threads.c: Fixup on the clean testdir

chapuni committed:
Rename threads.c to threads.test since it is no longer C source file.

chapuni committed:
llvm/test/tools/llvm-cov/Inputs: Avoid wildcards `rm -rf %t*.dir`

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/test

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.cpp

chapuni committed:
Use `[[#min(C,n)]]` for tests

chapuni committed:
llvm-cov: Introduce `--binary-counters`

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/pair

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/replace

chapuni committed:
s/replace/subst/

chapuni committed:
LLVMCoverage: Unify getCoverageForFile and getCoverageForFunction

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
getSwitchImplicitDefaultCounterPair

chapuni committed:
Merge branch 'users/chapuni/cov/single/pair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Fold either in switchcase

chapuni committed:
Merge branch 'users/chapuni/cov/single/test' into users/chapuni/cov/single/merge

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts.cpp

chapuni committed:
Update tests

chapuni committed:
Merge branches 'users/chapuni/cov/single/merge' and 'users/chapuni/cov/binary' into users/chapuni/cov/single/refactor-base

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test
	llvm/test/tools/llvm-cov/showLineExecutionCounts.test

chapuni committed:
New SingleByteCoverage

chapuni committed:
Merge branch 'users/chapuni/cov/single/refactor' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

chapuni committed:
threads.c => threads.test (following #113114)

chapuni committed:
clang/test: Add tests that we missed while I was updating.

chapuni committed:
Introduce CounterMappingRegion::isBranch(). NFC.

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp
	llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h

chapuni committed:
Merge branches 'users/chapuni/cov/single/test' and 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Update the test

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

ornata commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1400,14 +1375,12 @@ CodeGenFunction::EmitCXXForRangeStmt(const CXXForRangeStmt &S,
 
   if (ExitBlock != LoopExit.getBlock()) {
     EmitBlock(ExitBlock);
+    incrementProfileCounter(true, &S);
```
comment on boolean parameter for readability?

e.g.` /*var=*/true`

chapuni replied:
Done. See also #120930.

chapuni committed:
Update desc

chapuni committed:
s/Count1/BinaryCount/

chapuni committed:
Merge branch 'main' into users/chapuni/cov/binary

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test
	llvm/test/tools/llvm-cov/showLineExecutionCounts.test

chapuni committed:
Merge branch 'users/chapuni/cov/binary' into users/chapuni/cov/single/refactor

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

chapuni committed:
Introduce BranchCounterPair

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/pair

chapuni committed:
void verifyCounterMap() const

chapuni committed:
Introduce the dedicated class CounterPair instead of std::pair

chapuni committed:
Merge branches 'users/chapuni/cov/single/getpair' and 'users/chapuni/cov/single/pair' into users/chapuni/cov/single/nextcount-base

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CodeGenPGO.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Introduce {UseExecPath, UseSkipPath} instead of {false, true}

chapuni committed:
Catch up the merge

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Eliminate (rewind) "refactor" changes

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Fix wrong merge resolutions

chapuni committed:
Reorganize CoverageMapping::SingleByteCoverage

chapuni committed:
Prune commented-out line

chapuni committed:
[Coverage] Make `MCDCRecord::Folded` as `[false/true]` with BitVector. NFC.

chapuni committed:
[Coverage] MCDC: Move `findIndependencePairs` into `MCDCRecord`

chapuni committed:
Merge branches 'users/chapuni/cov/merge/mcdcfold' and 'users/chapuni/cov/merge/mov_ind' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
[Coverage] Sort `MCDCRecord::ExecVectors` order by Bitmap index

chapuni committed:
llvm-cov: Refactor CoverageSummaryInfo. NFC.

chapuni committed:
Merge remote-tracking branches 'origin/users/chapuni/cov/single/refactor' and 'origin/users/chapuni/cov/single/unify' into users/chapuni/cov/merge/forfile-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/refactor

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test
	llvm/test/tools/llvm-cov/showLineExecutionCounts.test
	llvm/tools/llvm-cov/CodeCoverage.cpp
	llvm/tools/llvm-cov/SourceCoverageView.h

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/refactor

chapuni committed:
Merge branch 'users/chapuni/cov/single/refactor' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/refactor' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/summaryinfo

chapuni committed:
Merge branch 'users/chapuni/cov/merge/summaryinfo' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Add a test

chapuni committed:
Prune redundant logic

chapuni committed:
Merge branch 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
Expand RHS of MapToExpand. This will prevent recursion.

chapuni committed:
Append an explanation in the comment

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mov_ind

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mcdcsort-base' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mov_ind' into users/chapuni/cov/single/unify

chapuni committed:
Revert "Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort-base"

This reverts commit c33e8987289f2775792977ee279c4e458f8091d5, reversing
changes made to ad6726dbfac5074c856f41adadb3b5bc14580311.

chapuni committed:
Rewind switch DefaultCase. (to #113112)

chapuni committed:
Enable addCounters

chapuni committed:
Get rid of structual bindings

chapuni committed:
Flatten with getBranchCounterPair(SkipCntForOld)

chapuni committed:
Merge branch 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

chapuni committed:
Dissolve assert

chapuni committed:
Update comments

chapuni committed:
Decorate the mock

chapuni committed:
Will be pruned after the migration of SingleByte.

chapuni committed:
Add comments

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/replace

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

chapuni committed:
Merge branch 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/getpair

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/pair

chapuni committed:
Merge branch 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/pair

chapuni committed:
Merge branch 'users/chapuni/cov/single/pair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
[MC/DC] Update CoverageMapping tests

To resolve the error, rename mcdc-error-nests.cpp -> mcdc-nested-expr.cpp at first.

- `func_condop`
  A corner case that contains close decisions.
- `func_expect`
  Uses `__builtin_expect`. (#124565)
- `func_lnot`
  Contains logical not(s) `!` among MC/DC binary operators. (#124563)

mcdc-single-cond.cpp is for #95336.

chapuni committed:
[MC/DC] Refactor MCDC::State::Decision. NFC.

Introduce `ID` and `InvalidID`. Then `DecisionByStmt` can have three
states.

* Not assigned if the Stmt(Expr) doesn't exist.
* When `DecisionByStmt[Expr]` exists:
  * Invalid and should be ignored if `ID == Invalid`.
  * Valid if `ID != Invalid`. Other member will be filled in the
    Mapper.

chapuni committed:
[MC/DC] Refactor MCDCCoverageBuilder. NFC.

- Get rid of the old `DecisionStack` and dissolve it into push/pop
  `CurCondIDs` in `VisitBin`, since `VisitBin` is recursive.

- Introduce the new `DecisionStack` with `DecisionState` to handle the
  current `Decision` in nested `Decision`s.
  - The stack has the sentinel that has `DecisionExpr = nullptr`.
  - Split out `checkDecisionRootOrPush` from `pushAndAssignIDs` for
    non-BinOp. It assigns `CondID` to `E` (instead of assignment LHS
    in `pushAndAssignIDs`).
  - The stack is manupilated at the top Decision operator in `VisitBin`.
    - The stack grows at the entrance of the Decision with the initial
      state.
    - In the same level in `VisitBin`, the stack is popped and the
      `Decision` record is emitted.
- Introduce `DecisionEndToSince` to sweep `MCDCBranch`es partially in
  `cancelDecision`.

chapuni committed:
[MC/DC] Prune MCDCLogOpStack and use CGF.isMCDCDecisionExpr

`MCDCLogOpStack` is used only for detection of the Decision root. It
can be detected with `MCDC::State::DecisionByStmt`.

chapuni committed:
[MC/DC] Create dedicated MCDCCondBitmapAddr for each Decision

MCDCCondBitmapAddr is moved from `CodeGenFunction` into `MCDCState`
and created for each Decision.

In `maybeCreateMCDCCondBitmap`, Allocate bitmaps for all valid
Decisions and emit them order by ID, to prevent nondeterminism.

chapuni committed:
[MC/DC] Handle __builtin_expect as if parenthses

Fixes #124565

chapuni committed:
[MC/DC] Enable usage of `!` among `&&` and `||`

In the current implementation, `!(a || b) && c` was not treated as one
Decision with three terms.

Fixes #124563

chapuni committed:
Update ReleaseNotes

chapuni committed:
isValid()

chapuni committed:
Remove mcdc-single-cond.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/tests

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Prune mcdc-single-cond.cpp

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CodeGenFunction.h

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CGStmt.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/condop

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
clangCodeGen: Follow #142155

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/covgen

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp (#172557)

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/bitmapaddr

Conflicts:
	clang/lib/CodeGen/CodeGenFunction.h (#142155)

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Update mcdc-export-json.test (in #159119)

chapuni committed:
Add comments

chapuni committed:
emitCounterSetOrIncrement: Simplify.

- s/`TheMap`/`TheCounterPair`/
- Change the type of `Counter` to `ValueOpt`.

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mcdcsort' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/loop

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/condop

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
fix: LoopFalse should be used.

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/expect

Conflicts:
	clang/include/clang/AST/IgnoreExpr.h
	clang/lib/CodeGen/CodeGenFunction.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/covgen

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/logopstack' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/expect' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covgen' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
[Coverage] Introduce the predicate `hasSkipCounter(S)`. NFC.

Also, rename s/getIsCounterPair/getExecSkipCounterExistence/.

chapuni committed:
Merge commit 'cead042fd7b57f7bb454d4ffe6f0413658a4b069' into users/chapuni/cov/single/loop

chapuni committed:
Use hasSkipCounter

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CGExprScalar.cpp
	clang/lib/CodeGen/CodeGenPGO.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/condop' into users/chapuni/cov/single/execskipexists

chapuni committed:
hasSkipCounter

chapuni committed:
Merge branch 'users/chapuni/cov/single/execskipexists' into users/chapuni/cov/single/loop

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.proftext
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.yaml
